File: Week6Haskell/tictactoe.hs
data Player = X | O | N
  deriving (Eq)

instance Show Player where
    show X = "X"
    show O = "O"
    show N = "_" 
    
type Board = [Player]

displayBoard :: Board -> String
displayBoard [] = []
displayBoard (p1:p2:p3:ps) = show p1 ++ show p2 ++ show p3 ++ "\n" ++ displayBoard ps

rowWinner :: Player -> Board -> Bool
rowWinner _ [] = False
rowWinner p (p1:p2:p3:ps)
 | (p == p1) && (p == p2) && (p == p3) = True
 | otherwise = rowWinner p ps

transpose :: Board -> Board
transpose [p1, p2, p3, p4, p5, p6, p7, p8, p9] = [p1, p4, p7, p2, p5, p8, p3, p6, p9]

diagWinner :: Player -> Board -> Bool
diagWinner p [p1,p2,p3,p4,p5,p6,p7,p8,p9] 
 | (p == p1) && (p == p5) && (p == p9) = True
 | (p == p3) && (p == p5) && (p == p7) = True
 | otherwise = False

winner :: Player -> Board -> Bool 
winner p b
  | rowWinner p b = True
  | rowWinner p (transpose b) = True
  | diagWinner p b = True
  | otherwise = False

updateBoard :: Int -> Player -> Board -> Board
updateBoard 0 player (b:bs) = [player] ++ bs
updateBoard idx player (b:bs) = [b] ++ updateBoard (idx - 1) player bs

gameLoop :: [Player] -> Board -> IO ()
gameLoop [] _ = print "Game over!"
gameLoop (p:ps) b = do 
  putStrLn (displayBoard b)
  putStrLn ("It's player " ++ show p ++ "'s turn")
  putStrLn "Enter a Move Index (1-9)"
  move <- getLine
  let moveInt = read move :: Int
  let b2 = updateBoard (moveInt - 1) p b
  if winner p b2 then do
    putStrLn (displayBoard b2)
    putStrLn ("Player " ++ show p ++ " wins")
  else 
    gameLoop ps b2

  

main :: IO ()
main = do
  putStrLn "Welcome to Tic Tac Toe!"
  let turns = [O,X,O,X,O,X,O,X,O]
  let board = [N,N,N,N,N,N,N,N,N]
  gameLoop turns board

-- [O,O,O,O,O,O,O,O,O]

File: Week8Prolog/tictactoe.pl
board([-,-,-,-,-,-,-,-,-]).
turns([o,x,o,x,o,x,o,x,o]).

% display_board(+List)
display_board([]).
display_board([H1,H2,H3|T]) :-
    write(H1), write(H2), writeln(H3),
    display_board(T).

% row_winner(+Player,+List)
row_winner(P,[P,P,P|_T]) :- !.
row_winner(P,[_H1,_H2,_H3|T]) :- row_winner(P,T), !.

% transpose_board(+List1, -List2)
transpose_board([A,B,C, D,E,F, G,H,I],[A,D,G, B,E,H, C,F,I]).

% diagWinner(+Player,+List)
diagWinner(P, [ P, _B, _C, _D,  P, _F, _G, _H,  P]) :- !.
diagWinner(P, [_A, _B,  P, _D,  P, _F, P, _H, _I]) :- !.

% winner(+Player,+List)
% ; is the 'or' operator..
winner(P,X) :-
    row_winner(P,X),! ;
    transpose_board(X,Y), row_winner(P,Y), ! ;
    diagWinner(P,X).

% update_board(+Player,+List1,-List2,+Move)
update_board(P,[_|T],[P|T],1) :- !.
update_board(P,[H1|T1],[H1|T2], Move) :-
    Move2 is Move - 1,
    update_board(P,T1,T2,Move2).

% check_win(+Player,+List,+List)
check_win(Player,Board,_Turns) :-
    winner(Player,Board),!,
    display_board(Board),
    write(Player), writeln(' wins!').

check_win(_Player,Board,Turns) :-
    game(Board,Turns).

% game(+List,+List)
game(Board,[]) :-
    display_board(Board),
    writeln('Draw!'), !.

game(Board,[Player | RemainingTurns]) :-
    display_board(Board),
    format("It's Player ~a's turn:", Player),nl,
    writeln('enter a move (1-9)'),
    read(Move),
    update_board(Player,Board,NewBoard,Move),
    check_win(Player,NewBoard,RemainingTurns).

% tictactoe
tictactoe :- 
    board(Board),
    turns(Turns),
    game(Board,Turns).

:- tictactoe.
